java-checks/src/main/java/org/sonar/java/checks/StandardFunctionalInterfaceCheck.java

```diff
@@ -0,0 +1,299 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherCollection;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.java.resolve.Flags;
+import org.sonar.java.resolve.JavaSymbol;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Symbol.MethodSymbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S1711")
+public class StandardFunctionalInterfaceCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatcherCollection OBJECT_METHODS = MethodMatcherCollection.create(
+    methodMatcherWithName("equals", "java.lang.Object"),
+    methodMatcherWithName("getClass"),
+    methodMatcherWithName("hashcode"),
+    methodMatcherWithName("notify"),
+    methodMatcherWithName("notifyAll"),
+    methodMatcherWithName("toString"),
+    methodMatcherWithName("wait"),
+    methodMatcherWithName("wait", "long"),
+    methodMatcherWithName("wait", "long", "int"));
+
+  private static final Set<String> STD_INTERFACE_NAMES = new HashSet<>();
+
+  private static final Map<Integer, List<FunctionalInterface>> STD_INTERFACE_BY_PARAMETER_COUNT = new HashMap<>();
+
+  static {
+    registerInterface("java.util.function.BiConsumer<T,U>", "void", "T", "U");
+    registerInterface("java.util.function.BiFunction<T,U,R>", "R", "T", "U");
+    registerInterface("java.util.function.BinaryOperator<T>", "T", "T", "T");
+    registerInterface("java.util.function.BiPredicate<T,U>", "boolean", "T", "U");
+    registerInterface("java.util.function.BooleanSupplier", "boolean");
+    registerInterface("java.util.function.Consumer<T>", "void", "T");
+    registerInterface("java.util.function.DoubleBinaryOperator", "double", "double", "double");
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define a constant instead of duplicating this literal "double" 16 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1192)